src/docmanager/manager.ts

```diff
@@ -321,6 +273,60 @@ class DocumentManager implements IDisposable {
     return registry.getWidgetFactory(widgetName);
   }
 
+  /**
+   * Creates a new document, or loads one from disk, depending on the `which` argument.
+   * If `which==='create'`, then it creates a new document. If `which==='open'`,
+   * then it loads the document from disk.
+   *
+   * The two cases differ in how the document context is handled, but the creation
+   * of the widget and launching of the kernel are identical.
+   */
+  private _createOrOpenDocument(which: string, path: string, widgetName='default', kernel?: Kernel.IModel): Widget {
```

Comment:
`which: 'open' | 'create'` would be more explicit and avoid the defensive check below.